app/src/main/java/com/github/gogetters/letsgo/activities/ChatActivity.kt

```diff
@@ -1,54 +1,151 @@
+        const val TEXT_FIELD = "Text"
+    }
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_chat)
 
-        entryText = findViewById(R.id.chat_editText_input)
+        // get the current user info
+        val authInstance = FirebaseAuth.getInstance().currentUser
+        if (authInstance != null && authInstance.displayName != null) {
+            userName = authInstance.displayName!!
+            userId = authInstance.uid
+        } else {
+            userName = _defaultUsername
+        }
 
-        val listView: ListView = findViewById(R.id.chat_listView_messages)
+        entryText = findViewById(R.id.chat_editText_input)
+        listView = findViewById(R.id.chat_listView_messages)
         adapter = MessageAdapter(this)
         listView.adapter = adapter
+
+        // define listener for message update (chat bubbles) in listview
+        realtimeUpdateListener()
+
     }
 
+    override fun onDestroy() {
+        super.onDestroy()
+        Database.removeMessagesListener(_testingChatId, messageListener)
+    }
 
+    /**
+     * Called when the ImageButton is clicked in the corresponding layout
+     */
     fun sendMessage(view: View?) {
+        // get the text message to send when the button is clicked
         val messageText: String = entryText.text.toString()
 
         if (messageText.isNotEmpty()) {
-            // TODO send message. For the moment it is just displayed.
 
+            // clear the text field
             entryText.text.clear()
-            // get username to display
-            var userName = DEFAULT_USERNAME
-            val authInstance = FirebaseAuth.getInstance().currentUser
-            if (authInstance != null && authInstance.displayName != null) {
-                userName = authInstance.displayName!!
-            }
-            // display the message without sending
-            val message = ChatMessage(messageText, true, Calendar.getInstance().time, userName)
+
+            // create the message by matching the Firebase data structure
+//            val newMessage = mapOf(
+//                NAME_FIELD to userName,
+//                TEXT_FIELD to messageText
+//            )
+            userId ?: return
+
+            // send the message to the database
+            Database.sendMessage(userId!!, _testingChatId, messageText, {
+                Log.i("INFO", "message sent")
+                Toast.makeText(this@ChatActivity, "Message Sent", Toast.LENGTH_SHORT).show()
+            }, { e ->
+                Log.e("ERROR", e.toString())
+            })
+
+        }
+    }
+
+    /**
+     * Listener for the message update. Updates the message bubbles.
+     */
+    private fun realtimeUpdateListener() {
+
+        messageListener = Database.addMessagesListener(_testingChatId) { snapshot ->
+            val messageData = snapshot.getValue<MessageData>()
+
+            messageData!!
```

Comment:
i get this is checking if it is null, but if it is going to throw an exception anyway is it necessary? Maybe just let it throw later

Reply:
I think it doesn't compile if I let it be a nullable version